trieste/space.py

```
@@ -480,6 +503,186 @@ def __eq__(self, other: object) -> bool:
+            if any(len(ts) == 0 for ts in category_names):
+                raise ValueError("Category name lists cannot be empty")
+            tags = [tuple(ts) for ts in category_names]
+
+        self._tags = tags
+
+        ranges = [tf.range(len(ts), dtype=dtype) for ts in tags]
+        meshgrid = tf.meshgrid(*ranges, indexing="ij")
+        points = (
+            tf.reshape(tf.stack(meshgrid, axis=-1), [-1, len(tags)]) if tags else tf.zeros([0, 0])
+        )
+
+        super().__init__(points)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"CategoricalSearchSpace({self._tags!r})"
+
+    @property
+    def has_bounds(self) -> bool:
+        return False
+
+    @property
+    def lower(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def upper(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def tags(self) -> Sequence[Sequence[str]]:
+        """The tags of the categories."""
+        return self._tags
+
+    @property
+    def one_hot_encoder(self) -> EncoderFunction:
+        """A one-hot encoder for the numerical indices."""
+
+        def encoder(x: TensorType) -> TensorType:
+            flat_x, unflatten = flatten_leading_dims(x)
+            if flat_x.shape[-1] != len(self.tags):
+                raise ValueError(
+                    "Invalid input for one-hot encoding: "
+                    f"expected {len(self.tags)} tags, got {flat_x.shape[-1]}"
+                )
+            columns = tf.split(flat_x, flat_x.shape[-1], axis=1)
+            encoders = [
+                tf.keras.layers.CategoryEncoding(num_tokens=len(ts), output_mode="one_hot")
+                for ts in self.tags
+            ]
+            encoded = tf.concat(
+                [encoder(column) for encoder, column in zip(encoders, columns)], axis=1
+            )
```

Given `x` is now a float type, do we need to do some checking/casting similar to `to_tags`? I'm not sure if `tf.keras.layers.CategoryEncoding` allows non-integer inputs.

---

It's only the underling self.points in GeneralDiscreteSearchSpace that are now floats. The one hot encoding uses self.tags directly so isn't affected. (And this is all tested.)

---

The `x` passed in by the user are the underlying indices, which can be float. We then call the encoder (after flattening) with these. From what I can see, the testing uses integer indices only, but I could have easily missed that. But in any case, it doesn't matter too much as I just tried out `tf.keras.layers.CategoryEncoding` with float types and it silently converts them into integers (in a weird way, see below). But perhaps we want to be explicit with our checking/assertions and not rely on tensorflow.

```
> e = tf.keras.layers.CategoryEncoding(4, output_mode="one_hot")

> e(tf.constant([3.0, 2.999999], dtype=tf.float32))
<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 1.],
       [0., 0., 1., 0.]], dtype=float32)>

> e(tf.constant([3.0, 2.9999999], dtype=tf.float32))
<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0., 0., 0., 1.],
       [0., 0., 0., 1.]], dtype=float32)>
```